otx/hpo/hpo_runner.py

```diff
@@ -0,0 +1,387 @@
+    """
+
+    def __init__(self, num_gpu_for_single_trial: int = 1, available_gpu: Optional[str] = None):
+        check_positive(num_gpu_for_single_trial, "num_gpu_for_single_trial")
+
+        self._num_gpu_for_single_trial = num_gpu_for_single_trial
+        self._available_gpu = self._set_available_gpu(available_gpu)
+        self._usage_status: Dict[Any, List] = {}
+
+    def _set_available_gpu(self, available_gpu: Optional[str] = None):
+        if available_gpu is None:
+            cuda_visible_devices = os.getenv("CUDA_VISIBLE_DEVICES")
+            if cuda_visible_devices is not None:
+                available_gpu_arr = self._transform_gpu_format_from_string_to_arr(cuda_visible_devices)
+            else:
+                num_gpus = torch.cuda.device_count()
+                available_gpu_arr = list(range(num_gpus))
+        else:
+            available_gpu_arr = self._transform_gpu_format_from_string_to_arr(available_gpu)
+
+        return available_gpu_arr
+
+    def _transform_gpu_format_from_string_to_arr(self, gpu: str):
+        for val in gpu.split(","):
+            if not val.isnumeric():
+                raise ValueError(
+                    "gpu format is wrong. " "gpu should only have numbers delimited by ','.\n" f"your value is {gpu}"
+                )
+        return [int(val) for val in gpu.split(",")]
+
+    def reserve_resource(self, trial_id: Any) -> Optional[Dict]:
+        """Reserve a resource under 'trial_id'.
+
+        Args:
+            trial_id (Any): Name of trial to reserve the resource.
+
+        Raises:
+            RuntimeError: If there is already resource reserved by 'trial_id', then raise an error.
+
+        Returns:
+            Optional[Dict]: Training environment to use.
+        """
+        if not self.have_available_resource():
+            return None
+        if trial_id in self._usage_status:
+            raise RuntimeError(f"{trial_id} already has reserved resource.")
+
+        resource = list(self._available_gpu[: self._num_gpu_for_single_trial])
+        self._available_gpu = self._available_gpu[self._num_gpu_for_single_trial :]
+
+        self._usage_status[trial_id] = resource
+        return {"CUDA_VISIBLE_DEVICES": ",".join([str(val) for val in resource])}
+
+    def release_resource(self, trial_id: Any):
+        """Release a resource under 'trial_id'.
+
+        Args:
+            trial_id (Any): Name of trial which uses the resource to release.
+        """
+        if trial_id not in self._usage_status:
+            logger.warning(f"{trial_id} trial don't use resource now.")
+        else:
+            self._available_gpu.extend(self._usage_status[trial_id])
+            del self._usage_status[trial_id]
+
+    def have_available_resource(self):
+        """Check that there is available resource."""
+        return len(self._available_gpu) >= self._num_gpu_for_single_trial
+
+
+def get_resource_manager(
```

Comment:
Even when we consider the GPU training, CPU, GPU managers are required. 
Maybe this function can automatically select the proper manager, users might want to use CPU, GPU manager for the GPU training case.

This comment also consider the further works. (auto-config by checking the HW resource)

Reply:
It also uses the CPU manager when user selects CPU manager even if GPU is available. Could you check again?

Reply:
When we assume that the GPU training environment, then users might want to control CPU and GPU at the same time. So, in this case, this function should return both `CPUResourceManager` and `GPUResourceManager`. 

However, this function couldn't return the both of them. is it intended? if so, it's okay

Reply:
My intention is that `GPUResourceManager` is used for training w/ GPU and `CPUResourceManager ` is used for training w/ CPU. For now, `CPUResourceManager` doesn't manages resources considering real CPU resources. It just gets how many trials will be run in parallel. So it's not meaningful to use both of them.